File: code/fptn-client/http/websocket_client.cpp
#include "websocket_client.h"

#include <fmt/format.h>
#include <hv/requests.h>


using namespace fptn::http;


WebSocketClient::WebSocketClient(
    const std::string& vpnServerIP, 
    int vpnServerPort,
    const std::string& tunInterfaceAddress,
    bool useSsl,
    const NewIPPacketCallback& newIPPktCallback 
)
    :
        vpnServerIP_(vpnServerIP), 
        vpnServerPort_(vpnServerPort),
        token_(""),
        tunInterfaceAddress_(tunInterfaceAddress),
        newIPPktCallback_(newIPPktCallback)
{
    reconn_setting_t reconn;
    reconn_setting_init(&reconn);
    reconn.min_delay = 1000;
    reconn.max_delay = 10000;
    reconn.delay_policy = 2;

    ws_.setReconnect(&reconn);
    ws_.onopen = std::bind(&WebSocketClient::onOpenHandle, this);
    ws_.onmessage = std::bind(&WebSocketClient::onMessageHandle, this, std::placeholders::_1);
    ws_.onclose = std::bind(&WebSocketClient::onCloseHandle, this);
}

bool WebSocketClient::login(const std::string& username, const std::string& password) noexcept
{
    const std::string url = fmt::format("https://{}:{}/api/v1/login", vpnServerIP_, vpnServerPort_);
    const std::string request = fmt::format(R"({{ "username": "{}", "password": "{}" }})", username, password);

    std::cerr << "URL>" << url << std::endl;
    std::cerr << request << std::endl;

    auto resp = requests::post(url.c_str(), request);
    if (resp != nullptr) {
        try {
            auto response = nlohmann::json::parse(resp->body);
            if (response.contains("access_token")) {
                token_ = response["access_token"];
                std::cerr << "Login successful." << std::endl;
                return true;
            } else {
                std::cerr << "Error: Access token not found in the response." << std::endl;
            }
        } catch (const nlohmann::json::parse_error& e) {
            std::cerr << "Error parsing JSON response: " << e.what() << std::endl;
        }
    } else {
        std::cerr << "Error: Request failed or response is null." << std::endl;
    }
    return false;
}

bool WebSocketClient::start() noexcept
{
    const std::string url = fmt::format("wss://{}:{}/fptn", vpnServerIP_, vpnServerPort_);

    http_headers headers = {
        {"Authorization", token_},
        {"ClientIP", tunInterfaceAddress_}
    };
    if (ws_.open(url.c_str(), headers) != 0) {
        std::cerr << "Failed to open WebSocket connection" << std::endl;
    }

    th_ = std::thread(&WebSocketClient::run, this);
    return th_.joinable();
}

bool WebSocketClient::stop() noexcept
{
    if (th_.joinable()) {
        ws_.stop();
        th_.join();
        return true;
    }
    return false;
}

void WebSocketClient::setNewIPPacketCallback(const NewIPPacketCallback& callback) noexcept
{
    newIPPktCallback_ = callback;
}

bool WebSocketClient::send(fptn::common::network::IPPacketPtr packet) noexcept
{
    std::vector<std::uint8_t> serializedPacket = packet->serialize();
    ws_.send((const char*)serializedPacket.data(), serializedPacket.size());
    return true;
}

void WebSocketClient::run() noexcept
{
    // pass
}

void WebSocketClient::onOpenHandle() noexcept
{
    std::cout << "WebSocket connection opened" << std::endl;
}

void WebSocketClient::onMessageHandle(const std::string& msg) noexcept
{
    auto packet = fptn::common::network::IPPacket::parse((const std::uint8_t*)msg.c_str(), msg.size());
    if (packet != nullptr && newIPPktCallback_) {
        newIPPktCallback_(std::move(packet));
    }
}

void WebSocketClient::onCloseHandle() noexcept
{
    std::cout << "WebSocket connection closed" << std::endl;
}

File: src/android/app/src/main/java/org/fptn/client/service/websocket/WebSocketClient.java
package org.fptn.client.service.websocket;

import android.util.Log;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.security.cert.CertificateException;
import java.security.cert.X509Certificate;

import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLSocketFactory;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;

import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.WebSocket;
import okhttp3.WebSocketListener;
import okio.ByteString;

public class WebSocketClient implements Runnable {
    private String host;
    private Integer port;
    private String username;
    private String password;

    WebSocketCallback callback;

    private String accessToken;

    WebSocket webSocket = null;
    private static final String LOG_TAG = "FptnClient";

    public WebSocketClient(String host, Integer port, String username, String password, WebSocketCallback callback) {
        Log.i(LOG_TAG, "onStartCommand+" + host + " " + port + " " + username + " " + password);
        this.host = host; //"185.215.187.165";
        this.port = port;
        this.username = username;
        this.password = password;
        this.accessToken = "";
        this.callback = callback;
    }

    public boolean send(byte[] msg)
    {
        if (webSocket != null) {
            webSocket.send(msg.toString());
            return true;
        }
        return false;
    }

    public boolean login()
    {
        try {
            JSONObject json = new JSONObject();
            json.put("username", username);
            json.put("password", password);

            RequestBody requestBody = RequestBody.create(
                    json.toString(),
                    MediaType.get("application/json")
            );
            String url = String.format("https://%s:%d/api/v1/login", host, port);

            Request request = new Request.Builder()
                    .url(url)
                    .post(requestBody)
                    .build();

            OkHttpClient client = getUnsafeOkHttpClient();
            Response response = client.newCall(request).execute();
            if (response.code() == 200) {
                String responseBody =  response.body().string();
                JSONObject jsonResponse = new JSONObject(responseBody);
                if (jsonResponse.has("access_token")) {
                    accessToken = jsonResponse.getString("access_token");
                    System.out.println("Login successful. " + accessToken);
                    return true;
                } else {
                    System.err.println("Error: Access token not found in the response.");
                }
            }
        } catch (JSONException err) {
            Log.i(LOG_TAG, "Error1: " + err.toString());
        } catch (IOException err) {
            Log.i(LOG_TAG, "Error2: " + err.toString());
        }
        return false;
    }

    @Override
    public void run() {
        if (login()) {
            OkHttpClient client = getUnsafeOkHttpClient();
            WebSocketListener websocketListener = getWebSocketListener();
            Request request = new Request.Builder()
                    .url(String.format("wss://%s:%d/fptn", host, port))
                    .addHeader("Authorization", "Bearer " + accessToken)
                    .addHeader("ClientIP", "10.10.10.1")
                    .build();

            WebSocket webSocket = client.newWebSocket(request, websocketListener);
            try {
                Thread.sleep(100000);  // Adjust time to keep the connection open
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            // Shut down the client
            System.err.println("start");
            client.dispatcher().executorService().shutdown();
            System.err.println("stop");
        }
    }

    private WebSocketListener getWebSocketListener(){
        return new WebSocketListener() {
            @Override
            public void onOpen(WebSocket webSocket, Response response) {
                System.out.println("WebSocket opened!" + response.code());
                // Send a message after the connection is opened
//                webSocket.send("Hello WebSocket!");
            }

            @Override
            public void onMessage(WebSocket webSocket, String text) {
                System.out.println("Message received: " + text);
                callback.onMessageReceived(text.getBytes());
            }

            @Override
            public void onMessage(WebSocket webSocket, ByteString bytes) {
                System.out.println("Message received: " + bytes.hex());
                callback.onMessageReceived(bytes.toByteArray());
            }

            @Override
            public void onClosing(WebSocket webSocket, int code, String reason) {
                System.out.println("WebSocket closing: " + reason);
            }

            @Override
            public void onClosed(WebSocket webSocket, int code, String reason) {
                System.out.println("WebSocket closed: " + reason);
            }

            @Override
            public void onFailure(WebSocket webSocket, Throwable t, Response response) {
                t.printStackTrace();
                System.out.println("WebSocket failure: " + t.getMessage());
            }
        };
    }

    public static OkHttpClient getUnsafeOkHttpClient() {
        try {
            // Create a trust manager that does not validate certificate chains
            final TrustManager[] trustAllCerts = new TrustManager[]{
                    new X509TrustManager() {
                        @Override
                        public void checkClientTrusted(X509Certificate[] chain, String authType) throws CertificateException {
                        }

                        @Override
                        public void checkServerTrusted(X509Certificate[] chain, String authType) throws CertificateException {
                        }

                        @Override
                        public X509Certificate[] getAcceptedIssuers() {
                            return new X509Certificate[]{};
                        }
                    }
            };

            // Install the all-trusting trust manager
            final SSLContext sslContext = SSLContext.getInstance("SSL");
            sslContext.init(null, trustAllCerts, new java.security.SecureRandom());

            // Create an SSL socket factory with our all-trusting manager
            final SSLSocketFactory sslSocketFactory = sslContext.getSocketFactory();

            OkHttpClient.Builder builder = new OkHttpClient.Builder();
            builder.sslSocketFactory(sslSocketFactory, (X509TrustManager) trustAllCerts[0]);
            builder.hostnameVerifier((hostname, session) -> true);

            OkHttpClient client = builder.build();
//            client.readTimeout(0,  TimeUnit.MILLISECONDS);
            return client;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private static final String CHROME_CIPHERS = "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,"
            + "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,"
            + "TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,"
            + "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,"
            + "TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,"
            + "TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256,"
            + "TLS_RSA_WITH_AES_128_GCM_SHA256,"
            + "TLS_RSA_WITH_AES_256_GCM_SHA384,"
            + "TLS_RSA_WITH_AES_128_CBC_SHA,"
            + "TLS_RSA_WITH_AES_256_CBC_SHA,"
            + "TLS_RSA_WITH_3DES_EDE_CBC_SHA";
}
